tensorflow_decision_forests/contrib/temporal_feature_processor/temporal_feature_processor/core/evaluator.py

```diff
@@ -14,24 +14,98 @@
 
 """Evaluator module."""
 
-from typing import Dict, List, Union
-
-import pandas as pd
+import pathlib
+from typing import Any, Dict, List, Literal, Set, Union
 
+from temporal_feature_processor.core import backends
 from temporal_feature_processor.core.data.event import Event
+from temporal_feature_processor.core.data.feature import Feature
+from temporal_feature_processor.core.operators import base
+from temporal_feature_processor.implementation.pandas.data import event as pandas_event
 
-Query = Union[Event, Dict[str, Event], List[Event]]
-
-Data = Union[pd.DataFrame, Dict[str, pd.DataFrame], List[pd.DataFrame]]
-
-Result = Data
+AvailableBackends = Literal[tuple(backends.BACKENDS.keys())]
+Data = Dict[Event, Union[str, pathlib.Path, pandas_event.PandasEvent]]
+Query = Union[Event, List[Event]]
 
 
-def Eval(
+def evaluate(
     query: Query,
-    data: Data,
-) -> Result:
+    input_data: Data,
+    backend: AvailableBackends = "pandas",
+) -> Dict[Event, Any]:
   """Evaluates a query on data."""
 
-  del query
-  del data
+  if isinstance(query, Event):
+    events_to_compute = [query]
+
+  elif isinstance(query, list):
+    events_to_compute = query
+
+  else:
+    # TODO: improve error message
+    raise TypeError(
+        f"schedule_graph query argument must be one of {Query}. Received {type(query)}."
+    )
+
+  # get features from events
+  features_to_compute = [
+      feature for event in events_to_compute for feature in event.features()
+  ]
+
+  # get backend
+  selected_backend = backends.BACKENDS[backend]
+  event = selected_backend["event"]
+  evaluate_schedule_fn = selected_backend["evaluate_schedule_fn"]
+  read_csv_fn = selected_backend["read_csv_fn"]
+
+  # calculate opeartor schedule
+  schedule = get_operator_schedule(features_to_compute)
+
+  # materialize input data. TODO: separate this logic
+  materialized_input_data = {
+      input_event:
+      input_event_spec if isinstance(input_event_spec, event) else read_csv_fn(
+          input_event_spec, input_event.sampling())
+      for input_event, input_event_spec in input_data.items()
+  }
+  # evaluate schedule
+  outputs = evaluate_schedule_fn(materialized_input_data, schedule)
+
+  return {event: outputs[event] for event in events_to_compute}
+
+
+def get_operator_schedule(query: Set[Feature]) -> List[base.Operator]:
```

Comment:
Can you add a description of what this function is expecting to do, and possible a one-line on how this is currently done.

For example, the signature is probably something along the lines of: Returns a list L of operators such that the input features (and input event? according to yesterday chat) of operator L[i] are all computed by anterior operators { L[parent_1], L[parent_2], ..., L[parent_n]} where parent_j < i for-all i.

I think this function is correct, but it seems it could loop for-ever in the case of a cyclic graph. Could it be protected from this case?

Reply:
Yes - the function assumes that the graph is acyclic. I'll add a docstring and code to prevent a possible never-ending loop.